File: src/main/java/org/gga/graph/sort/TopologicalSort.java
package org.gga.graph.sort;

import org.gga.graph.Edge;
import org.gga.graph.Graph;
import org.gga.graph.exceptions.NotADagException;
import org.gga.graph.search.dfs.AbstractDfsVisitor;

/**
 * @author mike.aizatsky
 */
public class TopologicalSort {
    /**
     * Returns sorted vector list
     */
    public static int[] sort(final Graph g) throws NotADagException {
        final int[] result = new int[g.V()];

        g.getDfs().depthFirstSearch(new AbstractDfsVisitor() {
            int pos = g.V() - 1;

            @Override
            public void backEdge(Edge e, Graph graph) {
                throw new NotADagException();
            }

            @Override
            public void finishVertex(int v, Graph graph) {
                result[pos--] = v;
            }
        });
        return result;
    }
}

File: src/main/scala/org/gga/graph/sort/TopologicalSort.scala
package org.gga.graph.sort

import org.gga.graph.Edge
import org.gga.graph.Graph
import org.gga.graph.exceptions.NotADagException
import org.gga.graph.search.dfs.AbstractDfsVisitor

/**
 * @author mike.aizatsky
 */
object TopologicalSort {
  /**
   * Returns array with sorted vertices numbers.
   */
  def sort(g: Graph): Array[Int] = {
    val result: Array[Int] = new Array[Int](g.V)

    g.getDfs.depthFirstSearch(new AbstractDfsVisitor {
      private var pos: Int = g.V - 1

      override def backEdge(e: Edge, graph: Graph) {
        throw new NotADagException
      }

      override def finishVertex(v: Int, graph: Graph) {
        result(pos) = v
        pos -= 1
      }
    })
    result
  }
}